### DISCUSSION TITLE
Error when trying to run simulation on ChargEval

### DISCUSSION THREAD

**Discussion by faisalanshory** [s-t-lab/ChargEval#3] [Q&A]
Hello, I made this discussion as a continuation of the last one on the link https://github.com/chintanp/ChargEval/discussions/19

Context:
I have successfully deployed Evidence and Review components on Amazon EC2 Server and trying to add new EVSEs to the database and then run the simulation

Expectations:
The entered data can be received by the database and the simulation process can be run:

Reality:
After I put new EVSEs and then try to run the simulation, the screen gets blurry (Image 1) and then when I checked the server I found an error as shown in Image 2.

Problems encountered based on the previous discussion:
After trying to find out the problem from Figure 2, it is known that there is a data query problem where the data entered and required for simulation does not enter the database. I have quite a bit of trouble in the debugging process because I still lack experience. Do you have any suggestions on what parts to improve?

Image 1:
![158845684-e95fbfe7-eef6-4ba5-89d4-1a65d14a8841](https://user-images.githubusercontent.com/57836379/159977532-6c16afe3-f35a-4334-b017-69cdbb1ed8bd.jpeg)

Image 2:
![158845692-2e900aac-2373-4337-a6d8-cd0498284c6e](https://user-images.githubusercontent.com/57836379/159977539-9d7f0fc3-6e0f-4794-b716-c159d128c484.jpeg)

**Comment by chintanp** [comment]
@faisalanshory thank you for moving the discussion to the lab repo. Unfortunately, without further information, it is hard for me to debug this. You can try running the query(s) in a pgAdmin type setup against the database directly, to see whether it works directly or not. If it does, then you know that the problem is not with the formed query, but rather with the way R Shiny is inserting the data (less of chances of this being the cause). Some queries may be part of a transaction and may need to be executed as such. 

Have you created an account on Auth0 and provided the necessary credentials? The analysis_request does need a user-id associated so that the reverse lookup can be used to find all the relevant simulation submissions for each user.

@zackAemmer do you have any suggestions about the error above?

**Reply by chintanp** [reply]
@faisalanshory As promised the public AMIs for EVIABM (`ami-088e4f1ffa9f91e81`) and TRIPGEN (`ami-0b4a06671564698a9`). They should be accessible in the US-West-2 (Oregon) region. 

You can use the `ubuntu` user to login and then do `sudo su -` to change to the root user. You can then `cd` to the `/home/test` folder to view the related files. 

Let me know if they are not accessible or if you have any questions.

**Reply by faisalanshory** [reply]
Thank you for your suggestion Mr @chintanp. But could you please elaborate on what kind of query I should check/run in pgAdmin? Do I write code manually to insert data into the "new_evse" table? For your information, I have previously tried to run "base_version.sql" and restore the table with pgdump. All of the starting databases have managed to get into pgAdmin.

Regarding auth0, I've successfully registered and used it in the evides app. I'm assuming that I've managed to use auth0 correctly because it managed to get into the interface of the application. To make it clear, I will check again the cloud database if my account is registered and added to the database. 

I will also try to log in to the public AMIs and view related folder in /home/test folder. I will let you know if I have succeeded. Can you please tell me what is the goal of doing this because I do not know what should I check by doing it?

**Reply by faisalanshory** [reply]
Hello Mr @chintanp. I've been trying to run the problematic query that was shown on my AWS server in Image 2. Here are the results of the debugging that has been done:

### 1. Failed to write data to the analysis_record table:
 
![1a sim_date_time not called](https://user-images.githubusercontent.com/57836379/160675177-d6427b78-e7a0-4f14-8aed-83ba298a4170.jpg)

![image](https://user-images.githubusercontent.com/57836379/160675245-06f412cf-9bb5-4df4-8984-61e987b53aa1.png)

The query result tells that the sim_date_time column, which is a not-null constraint, is not called by INSERT INTO when interactively inputting data through the website

When modifying the query and inserting the sim_date_time field with dummy data, the code executes successfully as shown: 
![image](https://user-images.githubusercontent.com/57836379/160675518-ecc2fa09-4dea-44bc-b774-57f7fc51d459.png)

Note: Although the analysis_id column is not called by the INSERT INTO query but its value is filled with the number 8 (I don't know 
where the number comes from)

**Problem Conclusion1:
The not-null constraint sim_date_time column is not filled when inputting data from the website**

### 2. Failed to write data to new_evses table:

![image](https://user-images.githubusercontent.com/57836379/160676160-6c38ddc6-a1e8-4053-8d6c-e4d4bdbd8cbf.png)

In the error query displayed by the AWS server, there are 24 columns was called by INSERT INTO but in the first record there are only 21 inputs even though in the second record there are 24 inputs

When modifying the query and deleting the 1st record, another error appears:
![image](https://user-images.githubusercontent.com/57836379/160676366-b4a8290c-2e8c-456c-b2e1-2981bcf5a99f.png)

The error states that the currval of sequence is not yet defined even though there are already records in the analysis_record table.

When modifying the query so that the call is made by calling the last_value function as in the code sample in the first problem, another error is obtained:
![image](https://user-images.githubusercontent.com/57836379/160676997-bd317b65-27e7-4414-a256-8fa151b6f12b.png)

The result of the query tells that there is an input_evse_id field, with a not-null constraint, that is not called by INSERT INTO when it processes data input interactively through the website.
![image](https://user-images.githubusercontent.com/57836379/160677076-38b409b2-7475-417d-8448-c02095b78b61.png)

Note that the new_evses table has 27 columns but only 24 columns are called by the INSERT INTO query

When modifying the query and entering the input_evse_id field with dummy data, the code executes successfully:
![image](https://user-images.githubusercontent.com/57836379/160677171-1d391310-96d3-42b4-a47e-960aa7462086.png)

**Problem Conclusion:
a. The query written from the 2 new EVSES data entered through the website shows inconsistencies. It is known that the INSERT INTO query calls 24 columns, but the first data record only includes 20 values while the second data record includes 24 values. The called column in record 1 may be (level2_fixed_parking_price, level2_var_parking_price_unit, level2_var_parking_price) or (level2_fixed_charging_price, level2_var_charging_price_unit, level2_var_charging_price) or (dcfc_fixed_parking_price_var_unit),
b. Sequence calls using the CURRVAL query fail to execute so they must use the last_value function
c. The INSERT INTO query only calls 24 out of 27 columns, where one of the 3 uncalled columns, namely input_evse_id , has a not-null constraint character.**

### 3. Failed to write data to analysis_params table:

![image](https://user-images.githubusercontent.com/57836379/160677448-fe2b5b62-da30-4028-93c2-2923573241c4.png)

The error states that the currval of sequence is not yet defined even though there are already records in the analysis_record table.

When modifying the query so that the call is made by calling the last_value function as in the code sample in the first problem, it was ran successfully:
![image](https://user-images.githubusercontent.com/57836379/160677621-7c4a1eb2-65f4-4c4b-978a-5983780b1b0b.png)

**Problem Conclusion:
Sequence calls using the CURRVAL query fail to execute so they must use the last_value function**

**Reply by faisalanshory** [reply]
After debugging and knowing the root cause of the error encountered, in what file I can set and modify the query based on the findings? Do you have any input on what I should do next?

**Reply by chintanp** [reply]
@faisalanshory great work trying to debug this. And apologies, it is not more straightforward. A couple of comments below: 

1. The `analysis_id` in the `analysis_record` column is auto-generated using the Postgres SERIAL functionality: https://www.postgresqltutorial.com/postgresql-serial/.
2. The `CURRVAL` works in a transaction. So when the `analysis_record` is inserted first and then the other queries are run in the same session, then the CURRVAL will exist. This is the preferred approach compared to the last_val approach since multiple users of the framework may be using the platform simultaneously and someone else may have entered a value in the `analysis_record` table which would be picked for the other queries instead of the CURRVAL. This provides more info about sequences in Postgres: http://www.neilconway.org/docs/sequences/. 
3. This is where the queries are formed in EVIDES: https://github.com/chintanp/ev_infrastructure_designer/blob/ef7d42a66c3b6802390d73e2b2ab01d0b67f6f79/R/mod_config.R#L574. You can look at the individual functions and see if there is a discrepancy between what the database expects and what the query provides. In general, all the "needed" columns are not-nullable, so columns like `input_evse_id`, for example, can be ignored and won't affect the rest of the analysis. 
4. I am not sure if this is properly documented, but you can use Python to launch sims as well and do not need to use EVIDES. https://github.com/s-t-lab/ChargEval/blob/master/scripts/launch_sims_baasecase.ipynb is an example and there are several others in the folder `scripts`. The idea here is, that if you were inputting the location of new_evses from some other source like a csv etc., you can use the Python API to directly insert them into the database - which launches the sims. This can be used to verify that your database is in order, before using EVIDES. To make sure the database is in order, check that you have all the migrations executed against the database, as in the folder: https://github.com/chintanp/ChargEval/tree/master/database/migrations 
5. The AMI IDs (after you import them into your account - not sure if that is needed to launch an instance from a public AMI) for the TRIPGEN and EVIABM need to be provided as part of the ENV file in the ChargEVal directory. I will upload an example of that shortly. The AMIs themselves have sample env files of their own, that you will have to populate, essentially pointing it to your database. 

BTW, please feel free to submit pull requests for when and if you spot a bug in the code or docs and I would be happy to incorporate it.

**Reply by chintanp** [reply]
Added a .env-sample file: https://github.com/chintanp/ChargEval/blob/master/.env-sample. Some context: 
1. ChargEVal was previously called "EVI-DSS", so you might find some references and names related to that in places. 
2. You can ignore the containers and variables related to the dockprom - they are primarily for monitoring and logging, i.e. below the line: https://github.com/chintanp/ChargEval/blob/9ae0c5846c707d9b28893a96ded717819e1b25d4/docker-compose-prod.yml#L214. They will be helpful in the long run, but won't be part of the simulation process and can be safely brought in eventually. 

Otherwise, the migrations folder that I have in my master repo is inline with the migrations folder on Github, i.e. at 13.1.1, so your `flyway_schema_history` table should have 53 migrations with the latest one up to this number. This is the first and most important thing to get right. The whole framework is dependent on the database, so being consistent their is crucial.

**Reply by faisalanshory** [reply]
Thanks for the explanation. I'll probably try the suggestions you've given and will let you know the results.

Regarding pull requests, I'll do it when I'm sure the changes I've made are the right ones. I'm afraid that I'll just break the code or give the wrong corrections

**Reply by faisalanshory** [reply]
Hello Mr @chintanp ,

I've tried a few things regarding the feedback you've given. Even so, the application still fails to enter data into the database through app where the application hangs with a slightly gray appearance as before.

I've done several experiments:
1. Applications that have been successfully deployed can be accessed in https://chargeval.my.id/evides/
2. The "currval" command is replaced with the "SELECT last_value FROM .." function in mod_config.R and code of sql dump. The consideration is that the application may not be used simultaneously by a lot of people (according to the explanation regarding the reason for using currval) and the query code that fails to be executed by the server also fails to run when written on pgAdmin manually. After changing the currval, the query that is printed by the new mod_config.R is still not successfully executed by the server. But when writing the new query to pgAdmin, the data is successfully entered the database and the simulation function can be executed.

I also attach what is printed by the server when entering data through the application before and after the code change in the following link: https://drive.google.com/drive/folders/1Ev5b2OiN4RhodcPtmCSRgLB-gep6C0C-?usp=sharing

Do you have any suggestions about what should I do to fix the problem? Thank you in advance.

**Reply by faisalanshory** [reply]
Hello Mr @chintanp,

I would like to let you know that thankfully the input query can be run and the simulation process can be called as can be seen in the picture.
![WhatsApp Image 2022-04-09 at 18 57 44](https://user-images.githubusercontent.com/57836379/162576817-f1a99dd7-4e99-4a95-9899-9b91d9879e82.jpeg)

The application can be run after changing the value in the user's sql query to:
"INSERT INTO user_details(user_id, user_name, email_id) VALUES
                                     ('{auth0_userid}', '{user_name}', '{user_email}')
                                     ON CONFLICT (user_id) DO UPDATE SET last_submit_date = NOW();"
         )

I want to confirm approximately how long the simulation process will run? And there will be a notification via email when the process has been successfully executed right?

Thank you

**Comment by chintanp** [comment]
Good job getting this to run. Seems like this was a change in the Auth0
interface.

You see if the instance in running in your AWS console.

You can ssh into the instance to see how they are doing. If you have
configured the AMIs I shared with you with the variables to your DB etc.
And used the AMI id in your NodeJS ChargEVal env variables.

The tripgen typically takes 5-10 mins. And EVIABM takes roughly 3-4 hours.
If it is taking longer than 10 hours then something is definitely wrong.
Can also depend on parameters you have set.

I would happy to continue debugging your setup. I will share the AMI of
chargEVal as well and you can use grafana to monitor the individual
simulations. Once the EVIABM finishes you should get an email.

Thanks

**Reply by faisalanshory** [reply]
Thank you Mr @chintanp, I have a few following questions regarding the simulation process:
1. When trying to log in to the tripgen server while running, a log appears attached to the following link: https://drive.google.com/file/d/1_Jd7W8PO0TH2NY7MkB5vy-D9irux97wC/view?usp=sharing
2. Monitoring on the tripgen server does not display any processes with 0.3% cpu usage
![WhatsApp Image 2022-04-10 at 13 02 27](https://user-images.githubusercontent.com/57836379/162629803-bc82c310-4d99-4ec0-a1d1-d1a35434bb4e.jpeg)

3. Regarding the image that you gave, what branch is used for chargeval and tripgen? Because the tripgen image you provided uses the dev branch, will the chargeval image also use the dev branch?

**Reply by chintanp** [reply]
@faisalanshory It seems this runner log is generated for an analysis_id 23. So you can run queries against the database to see if tripgen generated any EV trips. `select * from evtrip_scenarios where analysis_id = 23` should give you that. In this case, it seems the EV trips were not generated. Does the database have access from the tripgen instance? You can run this locally to make sure it works. You can manually run a tripgen process locally. When it is done in EC2, the following code is run: https://github.com/chintanp/wsdot_evse_sim_manager/blob/23b59fe6d6e11a0fbf28bcad7439e536a1c356ec/index.js#L196 . You can replicate these steps and see if and where the tripgen code is failing. Also, you can see it checks out the deployment branch of the code. So if this is set to prod in the env file, then the master branch of the repo is used for trip generation and so on. Once the trips are generated the instances are stopped and terminated. Then EVIABM instances are launched. So if this did not happen automatically, then it means, the trip generation did not work as expected.

**Reply by faisalanshory** [reply]
Hello Mr @chintanp,

After trying your feedback, I encountered the following error:

Error in data.frame(dist = trip_sp$sp_len_ferry2, dest_charger_L2 = dest_charger_L2,  : 
  arguments imply differing number of rows: 1, 0

The code is accessible through http://54.68.3.91:8787/

With the debugging results as in the picture:

![WhatsApp Image 2022-04-17 at 10 04 20 AM](https://user-images.githubusercontent.com/57836379/163713818-c1edb1a3-1053-457c-ac93-d21864d32b56.jpeg)
![WhatsApp Image 2022-04-17 at 10 15 33 AM](https://user-images.githubusercontent.com/57836379/163713821-69dc5968-eb2e-4b44-a744-4475f9f6efd6.jpeg)

Do you have any idea why this error happened?

Thank you

**Reply by faisalanshory** [reply]
After trying to look for the root problem, it seems that the value of this logical dest_charger variable has no value, that's why the length of the variable when checked is equal to 0.

I tried to fill the dest_charger variable with a value of False so that the variable has a value and length that is not equal to 0. 
![1](https://user-images.githubusercontent.com/57836379/163740370-600f1460-63fa-4445-a20e-4a15d999feda.jpg)

As a result, Tripgen successfully ran the simulation. Evtrip scenarios have been successfully input as shown in the image:
![2](https://user-images.githubusercontent.com/57836379/163740376-79d7f51f-f375-45cd-b751-0b7fbd2e15f0.jpg)

The Service Simulation manager successfully started the Eviabm server to perform further simulations:
![3](https://user-images.githubusercontent.com/57836379/163740388-6441d342-976c-4406-abb5-74909c3b4032.jpg)
But there is problem with evi-abm code that cause no script/app is running (automatically) on the server yet. I've tried running it manually, but the result is that the simulation is completed in a second even though based on the documentation it can take several hours to finish. I am still trying to find why does it happen. 
![WhatsApp Image 2022-04-18 at 10 20 08 AM](https://user-images.githubusercontent.com/57836379/163763070-fc8df6b8-3ab5-491d-bcca-9e47ec892f79.jpeg)

What do you think about this approach, is it correct? How should the value for the dest_charger variable be determined? And do you have any idea about the problem of evi-abm?


Thank you, Mr @chintanp

**Reply by chintanp** [reply]
@faisalanshory Please check your database if the destination chargers table is populated. It needs to be seeded by running the following function: https://github.com/chintanp/wsdot_evse_update_states/blob/e959ace59d69225b8235799a4979c11b9627365a/R/generate_dc.R#L12 . This can be run either from your local machine or ChargEVal main EC2 instance (not from the ephemeral tripgen instance). It will need the db credentials for the main db in the  .Renviron file. This table is then appended with additional destination chargers if placed during the simulation run by a call to `update_dc()` in runner.R here (https://github.com/chintanp/wsdot_evse_update_states/blob/e959ace59d69225b8235799a4979c11b9627365a/runner.R#L59). This is already done and you do not need to do anything there. 

Similarly, you can run the function `update_wa_bevs()` here: https://github.com/chintanp/wsdot_evse_update_states/blob/e959ace59d69225b8235799a4979c11b9627365a/R/update_bevs.R#L18 and `update_evses()` here: https://github.com/chintanp/wsdot_evse_update_states/blob/e959ace59d69225b8235799a4979c11b9627365a/R/update_evses.R#L13 to update the EVs and EVSEs in the state to the latest. 

Let me know if you have any questions on the above or if you face any errors.

**Reply by chintanp** [reply]
Just saw the EVI-ABM error. For this, I suggest again running GAMA locally. The GAMA software with UI can be downloaded from the website: https://gama-platform.org/download and the model is here: https://github.com/chintanp/evi-abm. Use some `write()` statements to find out which line is causing the error and we then try to debug it. This code will need an env file as described in the sample.
